server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileRef.java

```diff
@@ -0,0 +1,174 @@
+
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.sonar.api.resources.Languages;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.server.util.LanguageParamUtils;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static org.apache.commons.lang.StringUtils.isEmpty;
+import static org.sonarqube.ws.client.component.ComponentsWsParameters.PARAM_ORGANIZATION;
+
+/**
+ * Reference to a Quality profile as defined by requests to web services api/qualityprofiles.
+ * The two exclusive options to reference a profile are:
+ * <ul>
+ *   <li>by its id (to be deprecated)</li>
+ *   <li>by the tuple {organizationKey, language, name}</li>
+ * </ul>
+ */
+public class QProfileRef {
+
+  public static final String DEFAULT_PARAM_LANGUAGE = "language";
+  public static final String DEFAULT_PARAM_PROFILE_NAME = "profileName";
+  public static final String DEFAULT_PARAM_PROFILE_KEY = "profileKey";
+
+  private final String key;
+  private final String organizationKey;
+  private final String language;
+  private final String name;
+
+  private QProfileRef(@Nullable String key, @Nullable String organizationKey, @Nullable String language, @Nullable String name) {
+    this.key = key;
+
+    this.organizationKey = organizationKey;
+    this.language = language;
+    this.name = name;
+  }
+
+  /**
+   * @return {@code true} if key is defined and {@link #getKey()} can be called. If {@code false}, then
+   *   the couple {language, name} is defined and the methods {@link #getLanguage()}/{@link #getName()}
+   *   can be called.
+   */
+  public boolean hasKey() {
+    return this.key != null;
+  }
+
+  /**
+   * @return non-null key
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code true}
+   */
+  public String getKey() {
+    checkState(key != null, "Key is not present. Please call hasKey().");
+    return key;
+  }
+
+  /**
+   * @return key of organization. It is empty when {@link #hasKey()} is {@code true} or if {@link #hasKey()} is
+   * {@code false} and the default organization must be used.
+   */
+  public Optional<String> getOrganizationKey() {
+    return Optional.ofNullable(organizationKey);
+  }
+
+  /**
+   * @return non-null language
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code false}
+   */
+  public String getLanguage() {
+    checkState(language != null, "Language is not present. Please call hasKey().");
+    return language;
+  }
+
+  /**
+   * @return non-null name
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code false}
+   */
+  public String getName() {
+    checkState(name != null, "Name is not present. Please call hasKey().");
+    return name;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)